### PR TITLE
hw4

### PR DESCRIPTION
**Pull request by alkorr37**

### CODE REVIEW

File: Homework4/src/domain/Laptop.java [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    private static boolean isManufacturerValid(String manufacturer) {
+        if (manufacturer != null) {
+            manufacturer = manufacturer.toUpperCase();
+        }
+        try {
+
+            Manufacturer.valueOf(manufacturer);
+            return true;
+        } catch (IllegalArgumentException e) {
+            // NOP
```
Comment (vitalii-kliuiev) [inline Homework4/src/domain/Laptop.java]: WTF

File: Homework4/src/domain/Laptop.java [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+    public Laptop setModel() {
+        this.model = "Unknown";
+        return this;
+    }
+
+    public Laptop setPrice(double price) {
+        this.price = isPriceValid(price) ? price : null;
+        return this;
+    }
```
Comment (vitalii-kliuiev) [inline Homework4/src/domain/Laptop.java]: Good implementation of Bilder )

File: Homework4/src/domain/Laptop.java [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+                .setManufacturer(this.manufacturer)
+                .setModel(this.model)
+                .setId(this.id)
+                .setCount(this.count)
+
+                .setCpu(this.cpu)
+                .setHddType(this.hddType)
+                .setLcd(this.lcd)
+                .setOs(this.os)
+                .setRam(this.ram);
```
Comment (vitalii-kliuiev) [inline Homework4/src/domain/Laptop.java]: What will be if you have a reference type (not String) ?

File: Homework4/src/domain/Monitor.java [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+    private static boolean isCountValid(int count) {
+        return count >= 0;
+    }
+
+    private static boolean isManufacturerValid(String manufacturer) {
+        try {
+            Manufacturer.valueOf(manufacturer);
+            return true;
+        } catch (IllegalArgumentException e) {
+            // NOP
```
Comment (vitalii-kliuiev) [inline Homework4/src/domain/Monitor.java]: excellent copy-paste)

File: Homework4/src/domain/Monitor.java [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+    private static boolean isModelValid(String model) {
+        return model != null;
+    }
+
+    private static boolean isPriceValid(double price) {
+        return price >= 0;
+    }
+
+    private static boolean isIdValid(BigInteger id) {
+        return id.intValue() > 0;
```
Comment (vitalii-kliuiev) [inline Homework4/src/domain/Monitor.java]: NullPointerException

File: Homework4/src/utils/Checker.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
+package utils;
+
+import domain.Laptop;
+import domain.Monitor;
+import domain.Product;
+
+public class Checker {
```
Comment (vitalii-kliuiev) [inline Homework4/src/utils/Checker.java]: use Interface for Utils, I can create the instance using reflection, private constructor does not prevent it

File: Homework4/src/domain/Laptop.java (hunk 1 above)
Comment (alkorr37) [reply Homework4/src/domain/Laptop.java]: No OPeration :)

File: Homework4/src/domain/Laptop.java (hunk 3 above)
Comment (alkorr37) [reply Homework4/src/domain/Laptop.java]: ?

File: Homework4/src/domain/Monitor.java (hunk 4 above)
Comment (alkorr37) [reply Homework4/src/domain/Monitor.java]: 8-)

File: Homework4/src/domain/Monitor.java (hunk 5 above)
Comment (alkorr37) [reply Homework4/src/domain/Monitor.java]: yep, my bug )

File: Homework4/src/utils/Checker.java (hunk 6 above)
Comment (alkorr37) [reply Homework4/src/utils/Checker.java]: but it's not interface in terms of OOP :) Why i should prevent you to instantiate it, especially using reflection?)))

Comment (vitalii-kliuiev) [reply Homework4/src/utils/Checker.java]: it is best practice

**Pull request merged**